thunder/benchmarks/__init__.py

```diff
@@ -2877,6 +2878,99 @@ def fn(self) -> Callable:
         return model
 
 
+class MoEBenchmark(Benchmark, metaclass=UserFacingBenchmarkMeta):
+    @staticmethod
+    def fused_topk_native(
+        hidden_states: torch.Tensor,
+        gating_output: torch.Tensor,
+        topk: int,
+        renormalize: bool,
+    ):
+        assert hidden_states.shape[0] == gating_output.shape[0], "Number of tokens mismatch"
+        M, _ = hidden_states.shape
+        topk_weights = torch.empty(M, topk, dtype=torch.float32, device=hidden_states.device)
+        topk_ids = torch.empty(M, topk, dtype=torch.int32, device=hidden_states.device)
+        topk_weights = torch.nn.functional.softmax(gating_output.float(), dim=-1)
+        topk_weights, topk_ids = torch.topk(topk_weights, topk, dim=-1)
+        if renormalize:
+            topk_weights = topk_weights / topk_weights.sum(dim=-1, keepdim=True)
+        return topk_weights, topk_ids
+
+    @staticmethod
+    def fused_moe_def(
+        x,
+        w1,
+        w2,
+        input_gating,
+        topk,
+        use_fp8_w8a8=False,
+        w1_scale=None,
+        w2_scale=None,
+        a1_scale=None,
+        a2_scale=None,
+    ) -> torch.Tensor:
+        assert not use_fp8_w8a8, "Not supported"
+
+        topk_weights, topk_ids = MoEBenchmark.fused_topk_native(
+            hidden_states=x,
+            gating_output=input_gating,
+            topk=topk,
+            renormalize=True,
+        )
+        w13_weights = w1[topk_ids]
+        w1_weights, w3_weights = torch.chunk(w13_weights, 2, dim=2)
+        w2_weights = w2[topk_ids]
+        x1 = torch.einsum("ti,taoi -> tao", x, w1_weights)
+        x1 = torch.nn.functional.silu(x1)
+        x3 = torch.einsum("ti, taoi -> tao", x, w3_weights)
+        expert_outs = torch.einsum("tao, taio -> tai", (x1 * x3), w2_weights)
+        return torch.einsum("tai,ta -> ti", expert_outs, topk_weights.to(expert_outs.dtype))
+
+    def __init__(
+        self,
+        tp_size: int,
+        batch_size: int = 128,
+        use_fp8: bool = False,
```

Comment:
Let's not use default values for users of this benchmark definition to spell out every option.
```suggestion
        *,
        tp_size: int,
        batch_size: int,
        use_fp8: bool,
```